### PR TITLE
Native discord notifications 5.0.x

### PR DESCRIPTION
**Pull request by recursivetree**
This upgrades @warlof 's work from https://github.com/eveseat/notifications/pull/60 to seat 5. I'll additionally add support for pings on top of it.

It seems to be working, but testing all messages will be impossible.

### CODE REVIEW

**Comment by recursivetree** [comment]
While working on it, I decided to slightly refactor how notifications work. Mostly, every notification extends from a Abstract<Integraton Type>Message, e.g. AbstractDiscordMessage. This allowed me to move for example the `via($notifiable)` in there, since we don't have any notification with multiple vias.

**Comment by recursivetree** [comment]
This is ready for being reviewed. I still get a few random failures and bugs I have to fix, but the refactors are done.

Yes, it's a lot of refactors, the PR title isn't entirely accurate anymore.

File: src/resources/views/groups/edit.blade.php [hunk 1]
```diff
@@ -6,296 +6,365 @@
...
+                {{ csrf_field() }}
+                <input name="id" value="{{ $group->id }}" type="hidden">
+
+                <div class="box-body">
+                    <div class="form-group">
+                        <label for="mention">{{ trans_choice('notifications::notifications.mention', 1) }}</label>
+                        <select name="mention_type" id="mention" class="form-control">
+                            @foreach(config("notifications.mentions") as $mention_type=>$mention)
+                                <option value="{{ $mention_type }}">
+                                        {{ ucfirst($mention["type"]) }}: {{ trans($mention["name"]) }}
```
Comment (warlof) [inline src/resources/views/groups/edit.blade.php:95]:
```suggestion
                                        {{ ucfirst($mention["type"]) }}: {{ trans($mention["label"]) }}
```

**Review by warlof — COMMENTED** [review]
Seems you got some fun with that one - well designed :)

Question: why a trait rather than an abstract class for the mapGroups/router logic ?

Just a few code style things:
- I try to introduce some enhanced log format policy in order to easily track the source of the log entry, as a result, I've updated the logging lines
- avoid to use special characters in key like `@`

Logic:
Notifications handler is following this structure:
```
Thing_To_Notify
- label for user displayable item
- handlers for target channel format
- handlers.{platform}
```

This structure is allowing a single entry for a thing which need to be notified independantly of the target platform. Hence, you got a more compact drilldown - and when the platform is available, we can highlight it using icon.

Mention handler is following this structure:
```
Mention_Kind_For_Platform
- type to determine the used platform
- label for user displayable item
- creation_handler for specific mention kind platform setup
- message_adapter for specific platform format
```

Instead, couldn't apply the same logic for mentions than the one used for notifications ?

```
Mention_Kind
- label for user disaplyable item
- platforms for target platform
- platforms.{platform}
- platforms.{platform}.creation_handler for fully qualified method name handling platform mention custom settings
- platforms.{platform}.message_adapter for fully qualified method name handling mention formatting
```

Then, using this https://github.com/eveseat/notifications/blob/71cc11e82cc35211cc04533856bd441079e52d35/src/resources/views/groups/edit.blade.php#L329#L364 you could format drilldown output and append platforms icon near to each mention.

File: src/Config/notifications.alerts.php [hunk 2]
```diff
@@ -169,7 +182,6 @@
     'OwnershipTransferred' => [
         'label'   => 'notifications::alerts.ownership_transferred',
         'handlers' => [
-            'mail' => \Seat\Notifications\Notifications\Structures\Mail\OwnershipTransferred::class,
```
[the remark this answers is not in the document]
Comment (recursivetree) [reply src/Config/notifications.alerts.php:172]: There is no implementation for the notification in \Seat\Notifications\Notifications\Structures\Mail\

**Comment by recursivetree** [comment]
> Question: why a trait rather than an abstract class for the mapGroups/router logic ?
Seat core only triggers notifications from observers and just speaking for seat core, an abstract class would be better suited. However, for example in my plugins I trigger notifications from various places, including observers, jobs and controllers. Some of them already inherit from something. You can't inherit from two classes, but from two traits. We can change it to an abstract class, but then I'll just have a trait like this in my plugins.

Removing @ from keys is no problem, I was sort of expecting it :).

Just in general, is there a style guide to follow? I sometimes feel a bit bad when it's almost only style issues.

> Thing_To_Notify and This structure is allowing a single entry for a thing which need to be notified independently of the target platform. Hence, you got a more compact drilldown - and when the platform is available, we can highlight it using icon.

I don't get it to be honest. What is Thing_To_Notify? A notification? A mention? or a group?

I don't really it makes sense to have one mention kind that works on multiple platforms, because most platforms are fundamentally different in how they work. Some cases, like mentioning a user are probably common to most platforms, but even then we might need different steps to set the mention up(e.g. discord and slack user IDs are not the same). That means that there needs to be a platform selector anyways. 

I agree, drilldown output should be added, regardless of whether we change how mentions work internally.

File: src/Config/notifications.alerts.php (hunk 2 above)
[the remark this answers is not in the document]
Comment (warlof) [reply src/Config/notifications.alerts.php:172]: anw, make sense then

**Comment by warlof** [comment]
> > Question: why a trait rather than an abstract class for the mapGroups/router logic ?
> > Seat core only triggers notifications from observers and just speaking for seat core, an abstract class would be better suited. However, for example in my plugins I trigger notifications from various places, including observers, jobs and controllers. Some of them already inherit from something. You can't inherit from two classes, but from two traits. We can change it to an abstract class, but then I'll just have a trait like this in my plugins.
> 
> Removing @ from keys is no problem, I was sort of expecting it :).
> 
> Just in general, is there a style guide to follow? I sometimes feel a bit bad when it's almost only style issues.
> 

You Shouldn't, really, but for guidelines, you got :
- PSR-2, PSR-4 - most of PSR-12
- sneeker variable naming
- upper camel case class naming
- lower camel case method naming

Then, as a rules of tumbs, avoiding reserved keyword and special characters as variable/method/key name

> > Thing_To_Notify and This structure is allowing a single entry for a thing which need to be notified independently of the target platform. Hence, you got a more compact drilldown - and when the platform is available, we can highlight it using icon.
> 
> I don't get it to be honest. What is Thing_To_Notify? A notification? A mention? or a group?
> 
> I don't really it makes sense to have one mention kind that works on multiple platforms, because most platforms are fundamentally different in how they work. Some cases, like mentioning a user are probably common to most platforms, but even then we might need different steps to set the mention up(e.g. discord and slack user IDs are not the same). That means that there needs to be a platform selector anyways.
> 
> I agree, drilldown output should be added, regardless of whether we change how mentions work internally.

Thing_To_Notify in my saying was referring to a Notification. I was trying to providing more meaningful names according to current spelling. Your vision of Adapters is more accurate according to what does classes inside this namespace `Seat\Notifications\Notifications` - the way notifications are implemented inside SeAT - `/Slack/*`, `/Mail/*` and `/Discord/*` are more kind of message formatted than the notification itself.

When talking about notification, we should more talk about things which are behind the event triggering the message against a platform- which is tracked by Observers classes - and the message itself.

Let's talk about this differently :

**Notification Registration and Message Formatter Configuration Structure**

```php
'notification_identifier' => [
    'label' => // translation string - to be used for UI elements
    'handlers' => [ // array of available platform with its related message format handler
        'mail' => \FullyQualifiedClassName\Mail\Message::class,
        'slack' => \FullyQualifiedClassName\Slack\Message::class,
        'platform' => \FullyQualifiedClassName\Message::class,
    ],
]
```

**Current Mention Registration and Mention Formatter Configuration Structure**

```php
'mention_platform_identifiier' => [
    'type' => // mention platform
    'label' => // translation string - to be used for UI elements
    'creation_controller_method' => // \FullyQualifiedClassName@method handling the mention configuration for the platform
    'message_adapter' => // \FullyQualifiedClassName@method handling message format to use the mention on the platform
]
```

**Alternative Mention Registration and Mentioon Formatter Configuration Structure**

```php
'mention_identifier' => [
    'label' => // translation string - to be used for UI elements
    'configuration_handlers' => [
        'discord' => // \FullyQualifiedClassName@method handling the mention configuration for the Discord platform
        'slack' => // \FullyQualifiedClassName@method handling the mention configuration for the Slack platform
    ],
    'message_adapters' => [
        'discord' => // \FullyQualifiedClassName@method handling message format to use the mention for the Discord platform
        'slack' => // \FullyQualifiedClassName@method handling message format to use the mention for the Slack platform
    ]
]
```

Base on this, the upper configuration can mirror its own logic from the Notification Registration - reduce duplication and as a rules of tumb : if we got the platform key - then the platform is supported by the mention

Anyway, just ideas

**Comment by recursivetree** [comment]
I agree, what you suggest for the structure would be more in line with how the alerts are structured. 

What I wonder though is what are the advantages of having it unified. I think the current way is clear enough that anyone who want's to implement a new mention can do it, additionally, how many times is this part of the code touched? That leaves out that we already have a working implementation of the other way.

The other thing that's not clear to me is wether you just want to change the way data is represented in the backend, or wether it should be user-visible. With user-visible I mean that the user selects 'Ping User' instead of 'Ping User on Discord'. That approach might sound more user-friendly, but it brings a whole lot of issues. For example:
1. A group can have multiple integrations. Do you have to set up mentions for all of them?
2. You might want to ping @everyone on discord but not on slack, meaning we need a platform selector anyways.
3. What happens if you add a new integration to the group, especially combined with 1.

These issues sound difficult to solve.

When the new data layout only affects the backend for above reasons, I don't have anything against it, but I don't think it is worth it either.

I another thing I read between the lines of your suggestion is that you want to just store the mention_identifier without platform in the db, but how do we then associated the configuration data with it?

**Pull request merged**